db/filters/base.py

```diff
@@ -0,0 +1,381 @@
+    type: LeafPredicateType = static(LeafPredicateType.NOT_EMPTY)
+    name: str = static("Not empty")
+
+    def to_sa_filter(self):
+        return column(self.column) != None
+
+
+@frozen_dataclass
+class In(MultiParameter, Leaf):
+    type: LeafPredicateType = static(LeafPredicateType.IN)
+    name: str = static("In")
+
+    def to_sa_filter(self):
+        return column(self.column).in_(self.parameters)
+
+
+@frozen_dataclass
+class NotIn(MultiParameter, Leaf):
+    type: LeafPredicateType = static(LeafPredicateType.NOT_IN)
+    name: str = static("Not in")
+
+    def to_sa_filter(self):
+        return column(self.column).not_in(self.parameters)
+
+
+@frozen_dataclass
+class Not(SingleParameter, Branch):
+    type: BranchPredicateType = static(BranchPredicateType.NOT)
+    name: str = static("Not")
+
+    def to_sa_filter(self):
+        return not_(*self.get_child_sa_filters())
+
+
+@frozen_dataclass
+class And(MultiParameter, Branch):
+    type: BranchPredicateType = static(BranchPredicateType.AND)
+    name: str = static("And")
+
+    def to_sa_filter(self):
+        return and_(*self.get_child_sa_filters())
+
+
+@frozen_dataclass
+class Or(MultiParameter, Branch):
+    type: BranchPredicateType = static(BranchPredicateType.OR)
+    name: str = static("Or")
+
+    def to_sa_filter(self):
+        return or_(*self.get_child_sa_filters())
+
+
+def get_predicate_subclass_by_type_str(predicate_type_str: str) -> Union[Type[LeafPredicateType], Type[BranchPredicateType]]:
+    for subclass in all_predicates:
+        if subclass.type.value == predicate_type_str:
+            return subclass
+    raise UnknownPredicateType(predicate_type_str)
+
+
+class BadFilterFormat(SABadFilterFormat):
+    pass
+
+
+class UnknownPredicateType(BadFilterFormat):
+    pass
+
+class ReferencedColumnsDontExist(BadFilterFormat):
+    pass
+
+
+all_predicates = [
+    Equal,
+    NotEqual,
+    Greater,
+    GreaterOrEqual,
+    Lesser,
+    LesserOrEqual,
+    Empty,
+    NotEmpty,
+    In,
+    NotIn,
+    Not,
+    And,
+    Or,
+]
+
+
+def _not_empty(xs):
+    return len(xs) > 0
+
+
+def _all_items_unique(xs):
+    for item1 in xs:
+        times_seen = 0
+        for item2 in xs:
+            if item1 == item2:
+                times_seen += 1
+            # A non-duplicate will be seen only once.
+            if times_seen == 2:
+                return False
+    return True
+
+
+def assert_predicate_correct(predicate):
+    """Enforces constraints on predicate instances."""
+    try:
+        if isinstance(predicate, Leaf):
+            column = predicate.column
+            column_name_valid = column is not None and type(column) is str and column != ""
+            assert column_name_valid, f"Column name invalid: {column}. It must be a non-empty string."
+
+        if isinstance(predicate, SingleParameter):
+            parameter = predicate.parameter
+            assert parameter is not None, "A parameter must not be None."
```

Comment:
Can (Custom)Exceptions be raised instead of assertions error? assert seem to indicate an internal error(something that should not happen), but this exception seems to be a mistake from the client

Reply:
I'd disagree. I use an `assert` as an abstraction over "if not true then throw generic exception with this information". I then catch them and wrap them in something useful. This is pretty much a shortcut over using a custom helper like `assertNotBadFilterFormat(parameter is not None, "A parameter must not be None.")`.

Reply:
I am assuming these are like validation errors rather than something which shouldn't have occurred. If that is the case I think we should be using generic `Exception` or some other subclass of `Exception`. Assertion errors by definition are used when an unexpected state or situation occurs due to fault of a developer. `ValueError` seems to a much better options

Reply:
@silentninja thanks, I wasn't aware that `assert` calls expand to an `if __debug__:` branch 🤦 I was thinking of an `assert`ion as a general purpose assertion.